File: app/src/main/java/com/example/bunnix/database/supabase/storage/ProductStorage.kt
package com.example.bunnix.database.supabase.storage

import android.content.Context
import android.net.Uri
import com.example.bunnix.database.config.SupabaseConfig
import io.github.jan.supabase.storage.storage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

object ProductStorage {

    private val storage = SupabaseConfig.client.storage

    // UPLOAD PRODUCT IMAGE
    suspend fun uploadProductImage(
        context: Context,
        productId: String,
        imageUri: Uri,
        imageIndex: Int
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            val file = uriToFile(context, imageUri)
            val fileBytes = file.readBytes()
            val path = "$productId/$imageIndex.jpg"

            val bucket = storage.from(SupabaseConfig.Buckets.PRODUCT_IMAGES)

            // ✅ CORRECT: Using options builder
            bucket.upload(
                path = path,
                data = fileBytes,
                options = {
                    upsert = true
                }
            )

            val publicUrl = bucket.publicUrl(path)
            file.delete()

            Result.success(publicUrl)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // DELETE PRODUCT IMAGE
    suspend fun deleteProductImage(productId: String, imageIndex: Int): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val path = "$productId/$imageIndex.jpg"
            val bucket = storage.from(SupabaseConfig.Buckets.PRODUCT_IMAGES)
            bucket.delete(path)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private fun uriToFile(context: Context, uri: Uri): File {
        val inputStream = context.contentResolver.openInputStream(uri)
            ?: throw Exception("Cannot open URI")

        val tempFile = File(context.cacheDir, "temp_${System.currentTimeMillis()}.jpg")
        tempFile.outputStream().use { output ->
            inputStream.copyTo(output)
        }
        inputStream.close()

        return tempFile
    }
}

File: src/database/supabase/storage/productStorage.js
import { supabase, BUCKETS } from '../config';

/**
 * UPLOAD PRODUCT IMAGE
 */
export const uploadProductImage = async (productId, file, imageIndex) => {
  try {
    const fileName = `${productId}/${imageIndex}.jpg`;
    
    const { data, error } = await supabase.storage
      .from(BUCKETS.PRODUCT_IMAGES)
      .upload(fileName, file, {
        cacheControl: '3600',
        upsert: true
      });

    if (error) throw error;

    const { data: publicUrlData } = supabase.storage
      .from(BUCKETS.PRODUCT_IMAGES)
      .getPublicUrl(fileName);

    return { 
      success: true, 
      url: publicUrlData.publicUrl 
    };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

/**
 * UPLOAD MULTIPLE PRODUCT IMAGES
 */
export const uploadMultipleProductImages = async (productId, files) => {
  try {
    const uploadPromises = files.map((file, index) => 
      uploadProductImage(productId, file, index)
    );
    
    const results = await Promise.all(uploadPromises);
    
    const failedUploads = results.filter(r => !r.success);
    if (failedUploads.length > 0) {
      throw new Error('Some images failed to upload');
    }
    
    const urls = results.map(r => r.url);
    
    return { 
      success: true, 
      urls 
    };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

/**
 * DELETE PRODUCT IMAGE
 */
export const deleteProductImage = async (productId, imageIndex) => {
  try {
    const fileName = `${productId}/${imageIndex}.jpg`;
    
    const { error } = await supabase.storage
      .from(BUCKETS.PRODUCT_IMAGES)
      .remove([fileName]);

    if (error) throw error;

    return { success: true };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

/**
 * DELETE ALL PRODUCT IMAGES
 */
export const deleteAllProductImages = async (productId, imageCount) => {
  try {
    const fileNames = Array.from({ length: imageCount }, (_, i) => 
      `${productId}/${i}.jpg`
    );
    
    const { error } = await supabase.storage
      .from(BUCKETS.PRODUCT_IMAGES)
      .remove(fileNames);

    if (error) throw error;

    return { success: true };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};